underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsPositionedUnderFileInputStream.java

```diff
@@ -61,22 +80,50 @@ public int read(byte[] b) throws IOException {
     return read(b, 0, b.length);
   }
 
+  /**
+   * If there has been a number of sequential reads in a row,
+   * we move to regular buffered reads.
+   */
   @Override
   public int read(byte[] buffer, int offset, int length) throws IOException {
-    int bytesRead = ((FSDataInputStream) in).read(mPos, buffer, offset, length);
+    int bytesRead;
+    if (mSequentialReadCount >= SEQUENTIAL_READ_LIMIT) {
+      ((Seekable) in).seek(mPos);
```

Comment:
is this `seek` necessary or guaranteed to be a no-op if say I have 10 continuous reads as we reset `mSequentialReadCount` and `seek` after every 3 reads?

Reply:
I changed this part a bit so we don't do the noop seek. Basically we do the seek lazily if we are in sequential mode, and don't reset the counter until we seek or skip a large step.